clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -511,17 +511,18 @@ private static int parseAcks(String acksString) {
      * @return The amount of time we waited in ms
      */
     private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
-        // add topic to metadata topic list if it is not there already.
-        if (!this.metadata.containsTopic(topic))
-            this.metadata.add(topic);
-
+        // add topic to metadata topic list if it is not there already and reset expiry
+        this.metadata.add(topic);
         if (metadata.fetch().partitionsForTopic(topic) != null)
             return 0;
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            // Add the topic to metadata before requesting update to ensure that the topic is included
+            // in the metadata topic list since unknown topics are removed from the metadata
+            this.metadata.add(topic);
```

Comment:
@hachikuji The issue we had was with short lived REST applications that were creating a random topic, publishing and consuming from it and then deleting the topic. Every time this happens, there is unnecessary load and log spam on the REST proxy for 5 minutes. The proxy could avoid publishing to non-existent topics by checking if the topic exists, so fixing the issue with deletion of unused topics would help avoid log spam in this case.